### PR TITLE
Some QoL functions for the Prelude

### PR DESCRIPTION
**Pull request by srush**
When writing the tutorial there were a bunch of instances that I wanted to be able to use but were not defined in the prelude. In particular being able to take the `mean` or `prod` over one of the axes of a matrix.  This adds: 

* Lifted Mul and Floating instances
* Accessors for inner axis (something like `mean $ axis2 x`)
* `maybe` function 
* `argmax` function
* Pair instance for `Arb` (should others have Pair too?)

### CODE REVIEW

File: lib/prelude.dx [hunk 1]
```diff
@@ -537,6 +545,30 @@ def sq  [Mul a] (x:a) : a = x * x
...
+  tanh   = map tanh
+  floor  = map floor
+  ceil   = map ceil
+  round  = map round
+  sqrt   = map sqrt
+  pow    = \x y. for i. pow x.i y.i
+  lgamma = map lgamma
+
+def axis1 (x : a => b => c) : b => a => c = for j. for i. x.i.j
+def axis2 (x : a => b => c => d) : c => a => b => d = for k. for i. for j. x.i.j.k
```
Comment (dan-zheng) [inline lib/prelude.dx:569]: Rather than "axis", how about renaming these to something like "permute" or "transpose"? Naming ideas not fully polished: just `permute` would be ambiguous, so it'd have to be like `permute{1,2}` or `swapaxes{0to1|0to2}`.

Maybe new APIs can be unlocked when there's a typeclass for index sets. Seems like frst-class index sets would add extra expressive power for defining APIs like axis swapping.

I like the "axis reordering" functions from Swift for TensorFlow, [named `Tensor.transposed(permutation:)`](https://github.com/tensorflow/swift-apis/blob/b7148bc6f5e76bd55e5b11ebd5dd14e42c6ef5a3/Sources/TensorFlow/Operators/Basic.swift#L375-L426):

```swift
extension Tensor {
  /// Returns a transposed tensor, with dimensions permuted in the specified order.
  @inlinable
  @differentiable(wrt: self where Scalar: TensorFlowFloatingPoint)
  public func transposed(permutation: Tensor<Int32>) -> Tensor {
    _Raw.transpose(self, perm: permutation)
  }

  /// Returns a transposed tensor, with dimensions permuted in the specified order.
  @inlinable
  @differentiable(wrt: self where Scalar: TensorFlowFloatingPoint)
  public func transposed(permutation: [Int]) -> Tensor {
    _Raw.transpose(self, perm: permutation)
  }
}
```

**Review by dan-zheng — APPROVED** [review]
LGTM!

File: lib/prelude.dx [hunk 2]
```diff
@@ -559,8 +591,8 @@ def scan' (init:a) (body:n->a->a) : n=>a = snd $ scan init \i x. dup (body i x)
 def fsum (xs:n=>Float) : Float = yieldAccum \ref. for i. ref += xs i
 def sum  [Add v] (xs:n=>v) : v = reduce zero (+) xs
 def prod [Mul v] (xs:n=>v) : v = reduce one  (*) xs
-def mean (xs:n=>Float) : Float = sum xs / IToF (size n)
-def std (xs:n=>Float) : Float = sqrt $ mean (map sq xs) - sq (mean xs)
+def mean [VSpace v] (xs:n=>v) : v = sum xs / IToF (size n)
+def std [Mul v, VSpace v, Floating v] (xs:n=>v) : v = sqrt $ mean (map sq xs) - sq (mean xs)
```
Comment (dan-zheng) [inline lib/prelude.dx:595]: Curious how the extra typeclass indirection affects performance, versus the original functions specialized for `n=>Float`. Would be nice to retroactively benchmark when benchmarking infra is ready.

Comment (dougalm) [reply lib/prelude.dx:595]: Run-time performance should be identical, because all type classes get fully specialized. There's a small compile-time cost.

File: lib/prelude.dx [hunk 3]
```diff
@@ -537,6 +545,30 @@ def sq  [Mul a] (x:a) : a = x * x
 def abs [Add a, Ord a] (x:a) : a = select (x > zero) x (zero - x)
 def mod (x:Int) (y:Int) : Int = rem (y + rem x y) y
 
+instance [Floating a] Floating (n=>a)
```
Comment (dougalm) [inline lib/prelude.dx:548]: Ha, for some reason I never thought to do this. It's a little touch of numpy-style rank polymorphism. I can't see a reason not to allow it.

File: lib/prelude.dx [hunk 4]
```diff
@@ -537,6 +545,30 @@ def sq  [Mul a] (x:a) : a = x * x
...
+  cosh   = map cosh
+  tanh   = map tanh
+  floor  = map floor
+  ceil   = map ceil
+  round  = map round
+  sqrt   = map sqrt
+  pow    = \x y. for i. pow x.i y.i
+  lgamma = map lgamma
+
+def axis1 (x : a => b => c) : b => a => c = for j. for i. x.i.j
```
Comment (dougalm) [inline lib/prelude.dx:568]: It might be a good idea to use `view` instead of `for`, for more predictable performance. With `for`, there's a risk that we'll materialize the permuted table, which might not be what you want.

File: lib/prelude.dx [hunk 5]
```diff
@@ -1221,6 +1253,11 @@ instance Arbitrary Int32
 instance [Arbitrary a] Arbitrary (n=>a)
   arb = \key. for i. arb $ ixkey key i
 
+instance [Arbitrary a, Arbitrary b] Arbitrary (a & b)
```
Comment (dougalm) [inline lib/prelude.dx:1256]: Glad to see you're finding this type class useful!

**Comment by dougalm** [comment]
Thanks Sasha, this is great!

File: lib/prelude.dx (hunk 2 above)
Comment (dan-zheng) [reply lib/prelude.dx:595]: Awesome, thanks!

File: lib/prelude.dx [hunk 6]
```diff
@@ -537,6 +545,30 @@ def sq  [Mul a] (x:a) : a = x * x
...
+  tan    = map tan
+  sinh   = map sinh
+  cosh   = map cosh
+  tanh   = map tanh
+  floor  = map floor
+  ceil   = map ceil
+  round  = map round
+  sqrt   = map sqrt
+  pow    = \x y. for i. pow x.i y.i
+  lgamma = map lgamma
```
Comment (oxinabox) [inline lib/prelude.dx:566]: I am not sure these are a great idea
One of the nice things about Dex (IMO) is not having broadcasting. (except the `broadcast` function)
Neither numpy's implict broadcasting (`sin(xs)`), nor julias explict broadcasting (`sin.(xs)`).
because `for i. sin(xs.i)` is just not that much longer and it is that much more clear (including how it generalizes)

Also we will run into problems if we generalize this to matrixes since `sin` and other expodentials are defines on matrixes.

File: lib/prelude.dx [hunk 7]
```diff
@@ -1259,13 +1296,16 @@ def maximumBy [Ord o] (f:a->o) (xs:n=>a) : a =
 def minimum [Ord o] (xs:n=>o) : o = minimumBy id xs
 def maximum [Ord o] (xs:n=>o) : o = maximumBy id xs
 
-def argmin [Ord o] (xs:n=>o) : n =
+def argscan (comp:o->o->Bool) (xs:n=>o) : n =
```
Comment (oxinabox) [inline lib/prelude.dx:1299]: I misread this as `args can`

File: lib/prelude.dx (hunk 3 above)
Comment (dan-zheng) [reply lib/prelude.dx:548]: Woah @ the rank polymorphism relevation. Reminds me of typeclass instances for `(->)` in Haskell, e.g. ["How QuickCheck generates random functions"](https://kseo.github.io/posts/2016-12-14-how-quick-check-generate-random-functions.html):

```haskell
class CoArbitrary a where
  coarbitrary :: a -> Gen b -> Gen b

-- | Modifies a generator using an integer seed.
variant :: Int -> Gen a -> Gen a

instance CoArbitrary Bool where
  coarbitrary False = variant 0
  coarbitrary True = variant 1

instance CoArbitrary a => CoArbitrary (Maybe a) where
  coarbitrary Nothing  = variant 0
  coarbitrary (Just x) = variant 1 . coarbitrary x

-- Finally, an answer to the question: how does QuickCheck generate random functions?
instance (Coarbitrary a, Arbitrary b) => Arbitrary (a -> b) where
  arbitrary = promote (\a -> coarbitrary a arbitrary)
```

File: lib/prelude.dx (hunk 7 above)
Comment (dan-zheng) [reply lib/prelude.dx:1299]: Alternative name: `def argcompare`

File: lib/prelude.dx (hunk 1 above)
Comment (srush) [reply lib/prelude.dx:569]: I thought about something fancy like this. In practice though in Dex I've found I never have more than 3 axes, and 90% of the time I just want to reduce over an inner axis. The names reflect that this is just popping an inner axis.

Comment (dan-zheng) [reply lib/prelude.dx:569]: That makes sense! Not over-designing and over-genericizing for now makes sense.

File: lib/prelude.dx (hunk 4 above)
Comment (srush) [reply lib/prelude.dx:568]: Oh I didn't even know about `view`. That's good to know.

File: lib/prelude.dx (hunk 5 above)
Comment (srush) [reply lib/prelude.dx:1256]: These are great. I like the new splitKey as well. 

I had written something similar for neural network initialization. Although for NN's there are some strange hacks like scaling based on matrix size, so they might need their own initializers. (Dex doesn't have newtypes right?)

Comment (dougalm) [reply lib/prelude.dx:1256]: Dex doesn't have newtypes, but I think you get the same functionality with ordinary `data` declarations. The main benefit of Haskell's `newtype` is that it's guaranteed to be zero-cost at runtime but that's true of Dex's `data` wrappers anyway. Haskell's `newtype` also allows things like `coerce` and fancy versions of `deriving` but Dex doesn't have those yet.

File: lib/prelude.dx (hunk 4 above)
Comment (dougalm) [reply lib/prelude.dx:568]: Not surprising. We added it [less than two weeks ago!](https://github.com/google-research/dex-lang/commit/5be588fc14d981b57028b550934f9c7b1523498b)

File: lib/prelude.dx (hunk 6 above)
Comment (srush) [reply lib/prelude.dx:566]: Hmm, would be interesting to think about this more. I don't really see the downside of using Dex's type system to do broadcasting safely. 

I was thinking about a table `a => b => c => Float` and wanting to take the `std` over the b dimension. (As now works with `add`)

`for i. for k. std for j. x.i.j.k`
versus
`std $ axis1 x`
